src/aggregator/aggregatorHandler.js

```diff
@@ -0,0 +1,181 @@
+/**
+ * Copyright (c) 2019, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+/**
+ * src/handlerUtil.js
+ * Kafka handler utilities
+ */
+
+const logger = require('pino')();
+const stats = require('stats-lite');
+const persist = require('./persist').persist;
+const { aggregatorTimeout, expectedEmits } = require('../config').getConfig();
+
+/**
+ * Key = JSON.stringify({
+ *  updatedAt,
+ *  name,
+ * });
+ */
+const aggregateMap = new Map();
+
+const flush = async (key) => {
+  const {
+    reqStartTime,
+    jobStartTime,
+    publishCompletedAt,
+    numClientsEmittedTo,
+    emittedAt,
+    acknowledgedAt,
+  } = aggregateMap.get(key);
+
+  const queueTime = jobStartTime - reqStartTime;
+
+  const isPublished = publishCompletedAt ? true : false;
+
+  const publishLatency = publishCompletedAt - jobStartTime;
+
+  const avgSubscribeLatency = Array.isArray(emittedAt) ?
+    stats.mean(emittedAt.map(emitTime => emitTime - publishCompletedAt)) : null;
+
+  const numSubsMissed = Array.isArray(emittedAt) ? expectedEmits - emittedAt.length : expectedEmits;
+
+  const isSuccessfullyEmitted = numSubsMissed === 0 ? true : false;
+
+  const numClientsAcknowledged = Array.isArray(acknowledgedAt) ?
+    acknowledgedAt.length : 0;
+
+  const endToEndLatency = Array.isArray(acknowledgedAt) ? acknowledgedAt.map(acknowledgeTime =>
+    acknowledgeTime - jobStartTime) : [];
+
+  const avgEndToEndLatency = stats.mean(endToEndLatency);
+  const ninetyFifthPercentileEndToEndLatency = stats.percentile(endToEndLatency, 0.95);
+  const medianEndToEndLatency = stats.median(endToEndLatency);
+
+  const aggregatedVal = {
+    jobStartTime,
+    queueTime,
+    publishLatency,
+    avgSubscribeLatency,
+    numSubsMissed,
+    avgEndToEndLatency,
+    medianEndToEndLatency,
+    ninetyFifthPercentileEndToEndLatency,
+    isPublished,
+    isSuccessfullyEmitted,
+    numClientsAcknowledged,
+  };
+  const parsedKey = JSON.parse(key);
+  aggregateMap.delete(key);
+  persist(parsedKey, aggregatedVal);
+};
+
+/**
+ * message = {
+ *  type,
+ *  reqStartTime,
+ *  jobStartTime
+ * }
+*/
+const requestStartedHandler = (message, key) => {
+  aggregateMap.set(key, {
+    reqStartTime: message.reqStartTime,
+    jobStartTime: message.jobStartTime,
+  });
+  setTimeout(() => {
+    flush(key);
+  }, aggregatorTimeout);
```

Comment:
Would it be too much work to track the expected and received messages so we can flush as soon as we've received them all? That way we could flush right away most of the time, and the timeout would only be needed as a backup when something gets dropped.

Reply:
Don’t know if that’s necessary though, that would mean i’d have to probably emit an event when we receive everything and timeout would still exist. Seems like a lot of extra code?

Reply:
Yeah, it's not necessary, it would be an optional optimization. If you have some extra time it might be worth looking into, if not it's fine as-is.